module/zfs/dmu.c

```diff
@@ -536,9 +538,16 @@ dmu_buf_hold_array_by_dnode(dnode_t *dn, uint64_t offset, uint64_t length,
 		zio = zio_root(dn->dn_objset->os_spa, NULL, NULL,
 		    ZIO_FLAG_CANFAIL);
 	blkid = dbuf_whichblock(dn, 0, offset);
+	if ((flags & DMU_READ_NO_PREFETCH) == 0 &&
+	    DNODE_META_IS_CACHEABLE(dn) && length <= zfetch_array_rd_sz) {
+		zs = dmu_zfetch_prepare(&dn->dn_zfetch, blkid, nblks,
+		    read && DNODE_IS_CACHEABLE(dn), B_TRUE);
+	}
 	for (i = 0; i < nblks; i++) {
 		dmu_buf_impl_t *db = dbuf_hold(dn, blkid + i, tag);
 		if (db == NULL) {
+			if (zs)
+				dmu_zfetch_run(zs, missed, B_TRUE);
```

Comment:
Since this is an error path, I don't think performance really matters.  I'm guessing that it's part of the API that if you call dmu_zfetch_prepare() then you have to call dmu_zfetch_run() to release some resource.  Could you document that in the comment above dmu_zfetch_prepare()?

Reply:
Yes.  If dmu_zfetch_prepare() returns a stream pointer, dmu_zfetch_run() must be called for it.  Added a comment for it.